### DISCUSSION TITLE
Coordinate of the first reflection point

### DISCUSSION THREAD

**Discussion by weirdooo-18** [NVlabs/sionna#360] [General]
1) Is there any way to figure the coordinates of the location at which the first reflection of a ray is happening? (if i consider max_depth to be 1)?    
2) And...can I find the delay or any other parameters for the path till the reflection (i.e., for the first half of the path) and then for the 2nd half of the path?

**Comment by jhoydis** [comment]
Hi @weirdoo-18,

Yes you can easily access all of this information but it is not officially documented. 

The [`Paths`](https://nvlabs.github.io/sionna/_modules/sionna/rt/paths.html#Paths) object has the undocumented  property `vertices` which corresponds to the sequence of intersection points of the rays with the scene. Then there are the properties `sources` and `targets` which denote the start and end point of every path.

A source/target corresponds to an antenna in the scene. This is different from a transmitter or receiver which can each have multiple antennas.

You can compute the distance to the first reflection point by computing the distance between the first vertex and the source of a path.
This is done, e.g., in the function `_compute_directions_distances_delays_angles` in the file [`solver_paths.py`](https://github.com/NVlabs/sionna/blob/main/sionna/rt/solver_paths.py).

Hope that helps.

**Reply by weirdooo-18** [reply]
Thank You so much sir.

**Reply by PhilSoy** [reply]
I confirm the usefulness of having the sequence of interaction points together with:
- the identity of the object on which the path is interacting
- the kind of interaction (reflect/trans/scat/diff/)
- Possibly the complex amplitude of the path after each interaction (probably less useful)

**Reply by jhoydis** [reply]
Hi @PhilSoy,

This information is already available. The `Paths` object has a property `objects` which contains the list of indices of the intersected scene objects  or wedges along each paths. 

Then there is the property `types` which describes the type of path. 

The complex amplitude after each interaction does not exist. Please see [Eq. (26)](https://nvlabs.github.io/sionna/em_primer.html#equation-h-final) for a definition of the complex channel gain. For every interaction, there is a transition matrix. These are however not stored in the `Paths` object.

**Reply by weirdooo-18** [reply]
Hello Sir,
           Is there a way to import " _compute_directions_distances_delays_angles" directly in sionna? I tried using sionna.rt.solver_paths, but coudn't import the function. It'll be of great help if I could do this.

Thank you.

**Reply by jhoydis** [reply]
This function cannot be directly imported as it is a private method of the `SolverPaths` class. However, if you really want to use it, you can access it via a `Scene` instance as
```
scene._solver_paths._compute_directions_distances_delays_angles
```

**Comment by akanksha-7018** [comment]
Thank you so much for mentioning about vertices...I was looking for it from a long time. But I have a doubt. I have 3 paths coming up for a certain dimension of the transmitter antenna. And the type of the first path is LoS, the other two is reflected. But, why is it that we get vertices coordinates for LoS as well? While collecting the vertices for all paths, is it better to avoid the vertices for LoS? 
![Screenshot 2024-04-09 160609](https://github.com/NVlabs/sionna/assets/166504906/15f6f7ce-dfc4-43b9-80ce-135abb73de86)

**Reply by jhoydis** [reply]
Hi @akanksha-7018,

You can ignore the vertices for the LoS paths. You have a similar artifact for paths of different depths. 

The reason is that the shape of the vertices tensor is ``[max_depth, num_targets, num_sources, max_num_paths, 3]``. That means that whenever you have paths with different depths, there are useless entries in the vertices tensor.
The same holds whenever you have a different number of paths between the sources and targets in a scene (Sources and Targets correspond to the points between which the ray tracer computes propagation paths).

Please note that you are working here with undocumented properties that were not expected to be exposed to the user. However these properties are documented in the source code.

**Reply by akanksha-7018** [reply]
Thank you Sir.
